### PR TITLE
feat(filter-set): Add filterset resource

### PR DESCRIPTION
**Pull request by ofekisr**
### SUMMARY
this is the second milestone in the filter set feature.
it is about managing a scope for filter set either Dashboard/User
1. Dashboard scoped filter sets will be available for all dashboard permitted user and will be maintained with admins dashboard owners
2. User scoped filter sets will be private for each user separately

implementation:
a subset API of dashboard ` "/<dashboard_id>/filtersets"` for managing a dashboard

notes:
future milestone will deal with sharing and collaborating on filter-sets

### TEST PLAN
tests suites
create_api_tests.py 
delete_api_tests.py
get_api_tests.py
update_api_tests.py

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [X] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [X ] Migration is atomic, supports rollback & is backwards-compatible
  - [X] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [X] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by ChronicJpk — APPROVED** [review]

File: superset/dashboards/filter_sets/api.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class FilterSetRestApi(BaseSupersetModelRestApi):
+    datamodel = SQLAInterface(FilterSet)
+    resource_name = "dashboard"
+    class_permission_name = FILTER_SET_API_PERMISSIONS_NAME
+    allow_browser_login = True
+    csrf_exempt = True
```
Comment (dpgaspar) [inline superset/dashboards/filter_sets/api.py:47]: `csrf_exempt` should be false

File: superset/dashboards/filter_sets/api.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    add_exclude_columns = ['id', OWNER_OBJECT_FIELD, DASHBOARD_FIELD]
+    add_model_schema = FilterSetPostSchema()
+    edit_model_schema = FilterSetPutSchema()
+    edit_exclude_columns = ['id', OWNER_OBJECT_FIELD, DASHBOARD_FIELD]
+    list_columns = ['created_on', 'changed_on', 'created_by_fk', 'changed_by_fk', NAME_FIELD,
```
Comment (dpgaspar) [inline superset/dashboards/filter_sets/api.py:52]: we also need to add `method_permission_name = MODEL_API_RW_METHOD_PERMISSION_MAP`

File: superset/dashboards/filter_sets/api.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+    def _init_properties(self) -> None:
+        super(BaseSupersetModelRestApi, self)._init_properties()
+
+    @expose("/<dashboard_id>/filtersets", methods=["GET"])
+    @protect()
+    @safe
+    @permission_name("get")
+    @rison(get_list_schema)
+    @merge_response_func(ModelRestApi.merge_order_columns, API_ORDER_COLUMNS_RIS_KEY)
```
Comment (dpgaspar) [inline superset/dashboards/filter_sets/api.py:72]: we don't actually need these on superset

**Review by dpgaspar — COMMENTED** [review]
First pass, left some comments

File: superset/app.py [hunk 4]
```diff
@@ -523,6 +523,9 @@ def init_views(self) -> None:
                     icon="fa-cog",
                 )
             appbuilder.add_separator("Data")
+        from superset.dashboards.filter_sets.api import FilterSetRestApi
+
+        appbuilder.add_api(FilterSetRestApi)
```
Comment (amitmiran137) [inline superset/app.py:528]:
```suggestion
        if feature_flag_manager.is_feature_enabled("DASHBOARD_NATIVE_FILTERS_SET"):
                from superset.dashboards.filter_sets.api import FilterSetRestApi
        
                appbuilder.add_api(FilterSetRestApi)
```
Lets add it under the feature flag

File: superset/app.py [hunk 5]
```diff
@@ -549,6 +549,10 @@ def init_views(self) -> None:
             "Data", cond=lambda: bool(self.config["DRUID_IS_ACTIVE"])
         )
 
+        from superset.dashboards.filter_sets.api import FilterSetRestApi
```
Comment (dpgaspar) [inline superset/app.py:552]: nit: import is done upstream

File: superset/common/request_contexed_based.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+from superset import conf, security_manager
+
+if TYPE_CHECKING:
+    from flask_appbuilder.security.sqla.models import Role
+
+
+def get_user_roles() -> List[Role]:
+    if g.user.is_anonymous:
+        public_role = conf.get("AUTH_ROLE_PUBLIC")
+        return [security_manager.find_role(public_role)] if public_role else []
```
Comment (dpgaspar) [inline superset/common/request_contexed_based.py:32]: can you use `security_manager.get_public_role` ?

File: superset/common/request_contexed_based.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    return g.user.roles
+
+
+def is_user_admin() -> bool:
+    user_roles = [role.name.lower() for role in list(get_user_roles())]
```
Comment (dpgaspar) [inline superset/common/request_contexed_based.py:37]: theres also this config key `AUTH_ROLE_ADMIN`

File: superset/dashboards/filter_sets/commands/exceptions.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+class FilterSetNotFoundError(ObjectNotFoundError):
+    def __init__(
+        self, filterset_id: Optional[str] = None, exception: Optional[Exception] = None
+    ) -> None:
+        super().__init__("FilterSet", filterset_id, exception)
+
+
+class FilterSetCreateFailedError(CreateFailedError):
+    base_massage = 'CreateFilterSetCommand of dashboard "%s" failed: '
```
Comment (dpgaspar) [inline superset/dashboards/filter_sets/commands/exceptions.py:38]: typo: `base_message`

File: superset/dashboards/filter_sets/commands/exceptions.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+    def __init__(
+        self, dashboard_id: str, reason: str = "", exception: Optional[Exception] = None
+    ) -> None:
+        super().__init__((self.base_massage % dashboard_id) + reason, exception)
+
+
+class FilterSetUpdateFailedError(UpdateFailedError):
+    base_massage = 'UpdateFilterSetCommand of filter_set "%s" failed: '
```
Comment (dpgaspar) [inline superset/dashboards/filter_sets/commands/exceptions.py:47]: copy/paste typo

File: superset/dashboards/filter_sets/commands/exceptions.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+    def __init__(
+        self, filterset_id: str, reason: str = "", exception: Optional[Exception] = None
+    ) -> None:
+        super().__init__((self.base_massage % filterset_id) + reason, exception)
+
+
+class FilterSetDeleteFailedError(DeleteFailedError):
+    base_massage = 'DeleteFilterSetCommand of filter_set "%s" failed: '
```
Comment (dpgaspar) [inline superset/dashboards/filter_sets/commands/exceptions.py:56]: copy/paste typo

**Comment by suddjian** [comment]
/testenv up SUPERSET_FEATURE_DASHBOARD_NATIVE_FILTERS=true SUPERSET_FEATURE_DASHBOARD_NATIVE_FILTERS_SET=true

File: superset/migrations/versions/3ebe0993c770_filterset_table.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+def upgrade():
+    op.create_table(
+        "filter_sets",
+        sa.Column("created_on", sa.DateTime(), nullable=True),
+        sa.Column("changed_on", sa.DateTime(), nullable=True),
+        sa.Column("id", sa.Integer(), nullable=False),
+        sa.Column("name", sa.VARCHAR(500), nullable=False),
+        sa.Column("description", sa.Text(), nullable=True),
+        sa.Column("json_metadata", sa.JSON(), nullable=False),
```
Comment (suddjian) [inline superset/migrations/versions/3ebe0993c770_filterset_table.py:41]: We should avoid including the type as part of the column/variable name. Also, if possible can we find a more descriptive name than `metadata`? I'm not sure what all is contained in `metadata`. Perhaps `configuration` or `definition`?

Comment (amitmiran137) [reply superset/migrations/versions/3ebe0993c770_filterset_table.py:41]: Ok so a bit context here .
The purpose of this field is for storing the list of native-filters config as they are stored today under the Dashboard metadata.

About the field name , I'm ok with anything we decide .

Let's see what others like .
@villebro any thoughts?

File: tests/integration_tests/dashboards/superset_factory_util.py [hunk 12]
```diff
@@ -145,11 +160,14 @@ def create_datasource_table_to_db(
...
     name: Optional[str] = None,
     db_id: Optional[int] = None,
+    database: Optional[Database] = None,
     owners: Optional[List[User]] = None,
 ) -> SqlaTable:
-    name = name or random_str()
-    owners = owners or []
-    db_id = db_id or create_database_to_db(name=name + "_db").id
+    name = name if name is not None else random_str()
+    owners = owners if owners is not None else []
```
Comment (eschutho) [inline tests/integration_tests/dashboards/superset_factory_util.py:167]: I see a lot of these changes from an 'or 'statement to explicitly checking for None. It doesn't look like there are other falsy values that you would have, I'm curious what the advantage of this change is?

**Comment by eschutho** [comment]
Is there anything that we can test for on this PR in the virtual env?  I see that most of the changes are  in the backend, so not sure if anything is visibly changed in the app.

**Comment by amitmiran137** [comment]
Zero ui changed and will be in followup PRs which this is big already

File: superset/dashboards/filter_sets/commands/base.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    def run(self) -> Model:
+        pass
+
+    def validate(self) -> None:
+        self._validate_filterset_dashboard_exists()
+
+    def _validate_filterset_dashboard_exists(self) -> None:
+        self._dashboard = DashboardDAO.get_by_id_or_slug(str(self._dashboard_id))
+        if not self._dashboard:
+            raise DashboardNotFoundError()
```
Comment (villebro) [inline superset/dashboards/filter_sets/commands/base.py:59]: I don't understand why we need a private method `_validate_filterset_dashboard_exists()` that is only called once by a one-liner wrapper `validate()`?

File: superset/models/dashboard.py [hunk 14]
```diff
@@ -397,6 +427,11 @@ def get(cls, id_or_slug: str) -> Dashboard:
         qry = session.query(Dashboard).filter(id_or_slug_filter(id_or_slug))
         return qry.one_or_none()
 
+    def am_i_owner(self) -> bool:
+        if g.user is None or g.user.is_anonymous or not g.user.is_authenticated:
+            return False
+        return g.user.id in set(map(lambda user: user.id, self.owners))
```
Comment (villebro) [inline superset/models/dashboard.py:433]: IMO `is_actor_owner()` would be a more appropriate name for this method.

File: superset/migrations/versions/3ebe0993c770_filterset_table.py (hunk 11 above)
Comment (villebro) [reply superset/migrations/versions/3ebe0993c770_filterset_table.py:41]: I suggest we not use `sa.JSON()` here but rather go with `sa.Text()` as we have until now, as with `sa.JSON()` we're not sure how it behaves on various metadata engines. So let's just keep it `Text()` and assume we need to parse/format when deserializing/serializing.

File: superset/models/filter_set.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+config = app.config
+logger = logging.getLogger(__name__)
+
+
+class FilterSet(Model, AuditMixinNullable):
+    __tablename__ = "filter_sets"
+    id = Column(Integer, primary_key=True)
+    name = Column(String(500), nullable=False, unique=True)
+    description = Column(Text, nullable=True)
+    json_metadata = Column(JSON, nullable=False)
```
Comment (villebro) [inline superset/models/filter_set.py:40]: same here (change to `Text`)

**Review by villebro — COMMENTED** [review]
First pass comments

File: superset/dashboards/filter_sets/commands/base.py (hunk 13 above)
Comment (ofekisr) [reply superset/dashboards/filter_sets/commands/base.py:59]: validate is coming from the base... and could be consists several of validators 
if you see only "validate" you can understand what is validate that is why, but you can understand what is "validate filterset dashboard exists

File: superset/models/dashboard.py (hunk 14 above)
Comment (ofekisr) [reply superset/models/dashboard.py:433]: > IMO `is_actor_owner()` would be a more appropriate name for this method.

agree

File: superset/commands/exceptions.py [hunk 16]
```diff
@@ -31,6 +31,26 @@ def __repr__(self) -> str:
...
+        object_type: str,
+        object_id: Optional[str] = None,
+        exception: Optional[Exception] = None,
+    ) -> None:
+        super().__init__(
+            _(
+                self.message_format.format(
+                    object_type, '"%s" ' % object_id if object_id else ""
+                )
+            ),
```
[the remark this answers is not in the document]
Comment (ofekisr) [reply superset/commands/exceptions.py:49]: image

File: superset/migrations/versions/3ebe0993c770_filterset_table.py (hunk 11 above)
Comment (ofekisr) [reply superset/migrations/versions/3ebe0993c770_filterset_table.py]: Dashboard model uses JSON as well and it is very convenient to use... If I expect to Json object why not use JSON?
why do you use int and not everything as string? 
Please explain to us why JSON is bad practice? how are you sure Text is ok? If you trust sqlalchemy you should use their interface.

File: superset/dashboards/filter_sets/commands/base.py (hunk 13 above)
Comment (villebro) [reply superset/dashboards/filter_sets/commands/base.py:59]: I would recommend not introducing abstractions before they are needed. If `validate` grows out of proportion it totally makes sense to break out the separate validators (we have linting rules for this), but until then I suggest just keeping it all in `validate()` with a comment explaining what each step is about if it's really difficult to understand (probably not needed here).

**Review by villebro — COMMENTED** [review]
A few quick questions, will review/test more tomorrow

File: superset/dashboards/filter_sets/commands/base.py (hunk 13 above)
Comment (ofekisr) [reply superset/dashboards/filter_sets/commands/base.py]: As written In the Clean Code book "Always try to explain yourself in code." 
and you should not use comments.
but I removed the BaseCommand so abstraction was removed

**Review by villebro — APPROVED** [review]
LGTM, thanks for all the hard work! Looking forward to the next iteration!

**Review by villebro — CHANGES_REQUESTED** [review]
Oh, I just noticed the migration down revision needs to be updated!

**Review by villebro — APPROVED** [review]

**Comment by stevetracvc** [comment]
Crap, I did this myself already (although mine is way more simple/naive). Any idea when the UI might get merged in? My names apparently conflict with these and break Superset, so if it's going to be a while I'll rename all of my stuff, and I might be able to help a bit with the UI.

Thanks!

**Pull request merged**